sound/soc/sof/sof-audio.c

```diff
@@ -189,12 +190,25 @@ int sof_restore_pipelines(struct device *dev)
 		switch (swidget->id) {
 		case snd_soc_dapm_dai_in:
 		case snd_soc_dapm_dai_out:
+			ipc_size = sizeof(struct sof_ipc_comp_dai) +
+				   sizeof(struct sof_ipc_comp_new_ext);
+			comp = kzalloc(ipc_size, GFP_KERNEL);
+			if (!comp)
+				return -ENOMEM;
+
 			dai = swidget->private;
-			comp_dai = &dai->comp_dai;
-			ret = sof_ipc_tx_message(sdev->ipc,
-						 comp_dai->comp.hdr.cmd,
-						 comp_dai, sizeof(*comp_dai),
+			/* restore the struct sof_ipc_comp_dai */
+			memcpy(comp, &dai->comp_dai,
+			       sizeof(struct sof_ipc_comp_dai));
+
+			/* append extended data to the end of the component */
+			memcpy((u8 *)comp + sizeof(struct sof_ipc_comp_dai),
+			       &swidget->comp_ext, sizeof(swidget->comp_ext));
```

Comment:
@keyonjie you are always appending the extended data even if UUID is not present in topology ? Please only send the extended data iff UUID is present in topology i.e. ABI>= 3.17
You can still kmalloc the structure with extended data, but only add it's size to hdr.size and the offset if ABI  >= 3.17 **EDIT** - I mean topology ABI.

Reply:
@lgirdwood then we might need to check if the UUID is null before setting the hdr.size and the offset in the driver side.
There are 2 issues to me with doing this in the driver side:
1. We don't have ABI information from the tplg file, so we need to check if the UUID token is assigned component by component. Theoretically there is possibility that only some of the components are defined with UUID tokens while other components are not, though we should try to avoid this.
2. when more fields are added to the extended data struct, we will hit similar issue, the existance of the UUID doesn't means we have the new fields, should we add the whole extended data structure size to hdr.size always and keep the new fields being 0s when we not have the field?

Reply:
@keyonjie Hmm, we do have topology ABI information, right? -> sof_manifest() 

It is true having a mix of UUID and non-UUID is a likely mistake to hit in topology files, so it's a good question how we deal with this. I'd propose to select the IPC interface based on the topology ABI version:

 - with 3.16 and older, IPC size is BASE (no extdata sent)
 - if tplg is with ABI 3.17 or newer, extended data always added (it's an error if some component is missing uuid info)
 - IPC struct size is fixed for 3.17, IPC size is BASE+UUID 
 - if another extended field is added in e.g. 3.19, IPC size is bumped again to BASE+UUID+NEWEXT

... i.e. choose the "IPC variant" to use when topology ABI version is discovered.

Reply:
> @keyonjie Hmm, we do have topology ABI information, right? -> sof_manifest()

Oh yes, thanks for the correction, I didn't notice this. It was not stored anywhere but for one time check only, we can save it for the following widget_load()s.

> 
> It is true having a mix of UUID and non-UUID is a likely mistake to hit in topology files, so it's a good question how we deal with this. I'd propose to select the IPC interface based on the topology ABI version:
> 
> * with 3.16 and older, IPC size is BASE (no extdata sent)

Yes, we can do this easily with the saved topology ABI version. The issue is that if the topology ABI is 3.17 while the kernel ABI is 3.19, should we send IPC size BASE + UUID or BASE+UUID+NEWEXT (with all new ext fields 0s)? if it is former, we need to send different IPC size for every different topology ABI version, we might need to introduce a table (which need update each time the minor version bump). If it is the later, we will send new ext fields all 0s and the FW should be able to check and recognize it is because of an older topology and we might need some fallback solution but not just return errors.

> * if tplg is with ABI 3.17 or newer, extended data always added (it's an error if some component is missing uuid info)

Yes, as the uuid is token parsed and saved to swidget, I would just send extended data for all components blindly, if the token is missed for some component, the extended uuid data will be all 0s, and the uuid matching inside FW will return error then.

> * IPC struct size is fixed for 3.17, IPC size is BASE+UUID
> * if another extended field is added in e.g. 3.19, IPC size is bumped again to BASE+UUID+NEWEXT
> 
> ... i.e. choose the "IPC variant" to use when topology ABI version is discovered.

Not exactly, if the topology ABI is newer than the driver ABI, e.g. in the example you listed above, the older driver might not recognize the NEWEXT, so we can't send IPC size with BASE+UUID+NEWEXT in this case.

Reply:
@keyonjie Good points. I think we must have some additional logic in kernel to figure out which ABI level we can work at. E.g. if topology is at 3.17 and kernel at 3.19, practically we are limited to 3.17 feature set. By our current ABI guidelines (https://github.com/thesofproject/sof/pull/3143), kernel can still send 3.17 IPC commands with all 3.18/19 extensions filled with zeroes. I know @lgirdwood has rejected code in FW to check for nul UUID content, with the rationale that if the sender is at newer ABI level, it should not add the UUID extension unless there is valid data. I think we can go with this (and update the ABI rules). That means kernel has to have logic to select the ABI variant and not use ABI extensions which are not supported by the loaded topology file. I.e. in your above example, kernel should not send 3.19 IPC messages if topology is too old. More complex for kernel, but from system point of view, makes sense to put this complexity on the kernel side than in the more resource constrained firmware.

Reply:
@keyonjie Can you update on the status of this? My take is that UUID is always added after this PR is added, but the merged FW implementation will fallback to type based check if no UUID matches. Is this correct?

Reply:
> @keyonjie Can you update on the status of this? My take is that UUID is always added after this PR is added, but the merged FW implementation will fallback to type based check if no UUID matches. Is this correct?

@kv2019i actually I had implemented an alternate solution for this, it didn't use the tplg ABI version, instead, it checked if the UUID token existed (UUID non-0s), if yes, the UUID existed, we will send the extended data to FW, otherwise, the header.size will not include the extended data, see this here:
```
	/* only non-zero UUID is valid */
	if (memcmp(&swidget->comp_ext, nil_uuid, SOF_UUID_SIZE))
		total_size += sizeof(swidget->comp_ext);
```

Reply:
@keyonjie Ack, I missed that update. Then this can be marked as resolved.